// lavas-core-vue/dist/utils/workbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.WORKBOX_PATH = undefined;
exports.getWorkboxFiles = getWorkboxFiles;
exports.useWorkbox = useWorkbox;

var _path = require('path');

var _fsExtra = require('fs-extra');

var _workboxWebpackPlugin = require('workbox-webpack-plugin');

var _workboxWebpackPlugin2 = _interopRequireDefault(_workboxWebpackPlugin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const WORKBOX_PATH = exports.WORKBOX_PATH = require.resolve('workbox-sw');

/**
 * get workbox files
 *
 * @param {boolean} isProd is in production mode
 * @return {Array} files .js & .map
 */
/**
 * @file utils.workbox.js
 * @author lavas
 */
function getWorkboxFiles(isProd) {
    let filename = isProd ? (0, _path.basename)(WORKBOX_PATH) : (0, _path.basename)(WORKBOX_PATH).replace('prod', 'dev');
    return [filename, `${filename}.map`];
}

/**
 * use workbox-webpack-plugin
 *
 * @param {Object} webpackConfig webpack config
 * @param {Object} lavasConfig lavas config
 */
function useWorkbox(webpackConfig, lavasConfig) {
    let {
        buildVersion,
        build: { publicPath, ssr },
        globals,
        router: { base = '/' },
        serviceWorker: workboxConfig
    } = lavasConfig;
    let { swSrc, appshellUrl, appshellUrls } = workboxConfig;

    if (base !== '/' && !base.endsWith('/')) {
        base += '/';
    }

    // service-worker provided by user
    let serviceWorkerContent = (0, _fsExtra.readFileSync)(swSrc);

    // import workbox-sw
    let importWorkboxClause = `importScripts('${publicPath}static/js/${getWorkboxFiles(true)[0]}');`;
    serviceWorkerContent = importWorkboxClause + serviceWorkerContent;

    // register navigation in the end
    let registerNavigationClause;
    if (ssr) {

        // be compatible with previous version
        if (appshellUrls && appshellUrls.length) {
            appshellUrl = appshellUrls[0];
        }

        // add build version to templatedUrls and set register clause
        if (appshellUrl) {
            // add router base
            if (base !== '/') {
                if (appshellUrl.startsWith('/')) {
                    appshellUrl = appshellUrl.substring(1, appshellUrl.length);
                }
                appshellUrl = base + appshellUrl;
            }

            workboxConfig.templatedUrls = {
                [appshellUrl]: `${buildVersion}`
            };

            registerNavigationClause = `workboxSW.router.registerNavigationRoute('${appshellUrl}');`;
        }
    } else {
        // https://github.com/lavas-project/lavas/issues/128
        registerNavigationClause = `workboxSW.router.registerNavigationRoute('${publicPath}index.html');`;
    }

    if (!/workboxSW\.router\.registerNavigationRoute/.test(serviceWorkerContent)) {
        if (/workboxSW\.precache\(\[\]\);/.test(serviceWorkerContent)) {
            serviceWorkerContent = serviceWorkerContent.replace(/workboxSW\.precache\(\[\]\);/, `workboxSW.precache([]);\n${registerNavigationClause}\n`);
        } else {
            serviceWorkerContent += registerNavigationClause;
        }
    }

    // write new service worker in .lavas/sw.js
    let tempSwSrc = (0, _path.join)(globals.rootDir, './.lavas', 'sw-temp.js');
    (0, _fsExtra.writeFileSync)(tempSwSrc, serviceWorkerContent, 'utf8');
    workboxConfig.swSrc = tempSwSrc;

    // use workbox-webpack-plugin@2.x
    webpackConfig.plugin('workbox').use(_workboxWebpackPlugin2.default, [workboxConfig]);
}
